File: typescript/packages/core/tests/read-only-tool-view.test.ts
/**
 * ReadOnlyToolView — the SelfVerifying eval-phase read-only tool filter
 * (spore-core SC-30, issue #153).
 *
 * Mirrors the Rust unit test
 * (`read_only_tool_view_filters_to_readonly_allowlist`): an inner registry
 * advertising read + write + exec tools is wrapped in a view restricted to the
 * read-only allow-list. The view advertises only the intersection, dispatches
 * allow-listed calls through to the inner registry, and blocks the rest with a
 * recoverable error WITHOUT reaching the inner registry.
 */

import { describe, expect, it } from "vitest";

import {
  toolRegistry,
  type ToolCall,
  type ToolOutput,
  type ToolRegistry,
  type ToolSchema,
} from "../src/index.js";

const { READONLY_EVAL_TOOL_NAMES, ReadOnlyToolView } = toolRegistry;

// Inner registry advertising read + write + exec tools; records every dispatch.
class InnerRegistry implements ToolRegistry {
  readonly dispatched: string[] = [];

  async dispatch(call: ToolCall): Promise<ToolOutput> {
    this.dispatched.push(call.name);
    return { kind: "success", content: "ok", truncated: false };
  }

  isAlwaysHalt(_toolName: string): boolean {
    return false;
  }

  schemas(): ToolSchema[] {
    return ["read_file", "write_file", "bash"].map((name) => ({
      name,
      description: "",
      input_schema: { type: "object" },
    }));
  }
}

function call(id: string, name: string): ToolCall {
  return { id, name, input: {} };
}

describe("ReadOnlyToolView (SC-30, #153)", () => {
  it("advertises only the read-only intersection of the inner catalogue", () => {
    const inner = new InnerRegistry();
    const view = new ReadOnlyToolView(inner, READONLY_EVAL_TOOL_NAMES);

    const names = view.schemas().map((s) => s.name);
    expect(names).toContain("read_file");
    expect(names).not.toContain("write_file");
    expect(names).not.toContain("bash");
  });

  it("dispatches an allow-listed tool through to the inner registry", async () => {
    const inner = new InnerRegistry();
    const view = new ReadOnlyToolView(inner, READONLY_EVAL_TOOL_NAMES);

    const out = await view.dispatch(call("1", "read_file"));
    expect(out.kind).toBe("success");
    expect(inner.dispatched).toEqual(["read_file"]);
  });

  it("blocks a non-allow-listed tool with a recoverable error, never reaching inner", async () => {
    const inner = new InnerRegistry();
    const view = new ReadOnlyToolView(inner, READONLY_EVAL_TOOL_NAMES);

    const out = await view.dispatch(call("2", "write_file"));
    expect(out.kind).toBe("error");
    if (out.kind === "error") {
      expect(out.recoverable).toBe(true);
      expect(out.message).toContain("read-only evaluate phase");
    }
    // The blocked call never reached the inner registry.
    expect(inner.dispatched).toEqual([]);
  });

  it("delegates isAlwaysHalt to the inner registry", () => {
    const inner = new InnerRegistry();
    const view = new ReadOnlyToolView(inner, READONLY_EVAL_TOOL_NAMES);
    expect(view.isAlwaysHalt("read_file")).toBe(false);
  });
});

File: go/spore-core/read_only_tool_view_test.go
package sporecore

import (
	"context"
	"encoding/json"
	"testing"
)

// rovInnerRegistry is a ToolRegistry test double for the SC-30 ReadOnlyToolView
// test. It advertises a fixed read + write + exec schema set and records every
// tool name dispatched so the view's pass-through / block behavior is
// observable. Only the ToolRegistry methods exercised by ReadOnlyToolView are
// meaningful; the rest satisfy the interface.
type rovInnerRegistry struct {
	dispatched []string
}

func (r *rovInnerRegistry) Register(_ Tool, _ RegistryToolSchema) error { return nil }
func (r *rovInnerRegistry) RegisterSet(_ ToolSet) error                 { return nil }

func (r *rovInnerRegistry) ActiveSchemas(_ *TaskPhase) []RegistryToolSchema {
	mk := func(name string) RegistryToolSchema {
		return RegistryToolSchema{
			Name:       name,
			Parameters: json.RawMessage(`{"type":"object"}`),
		}
	}
	// read_file is in the read-only allow-list; write_file + bash are not.
	return []RegistryToolSchema{mk("read_file"), mk("write_file"), mk("bash")}
}

func (r *rovInnerRegistry) Dispatch(
	_ context.Context,
	call ToolCall,
	_ SandboxProvider,
) (HarnessToolResult, error) {
	r.dispatched = append(r.dispatched, call.Name)
	return HarnessToolResult{CallID: call.ID, Output: NewToolOutputSuccess("ok")}, nil
}

func (r *rovInnerRegistry) DispatchAll(
	ctx context.Context,
	calls []ToolCall,
	sandbox SandboxProvider,
) []DispatchOutcome {
	out := make([]DispatchOutcome, len(calls))
	for i, c := range calls {
		res, _ := r.Dispatch(ctx, c, sandbox)
		out[i].Result = res
	}
	return out
}

func (r *rovInnerRegistry) HasSubagentTools() bool     { return false }
func (r *rovInnerRegistry) IsAlwaysHalt(_ string) bool { return false }

var _ ToolRegistry = (*rovInnerRegistry)(nil)

// TestReadOnlyToolViewFiltersToReadonlyAllowlist mirrors the Rust unit test
// (SC-30): the eval-phase read-only view advertises + dispatches only the
// INTERSECTION of the wrapped catalogue with the read-only allow-list.
func TestReadOnlyToolViewFiltersToReadonlyAllowlist(t *testing.T) {
	inner := &rovInnerRegistry{}
	view := NewReadOnlyToolView(inner, readonlyEvalAllowSet())

	// Schemas: only read_file survives the intersection; write_file / bash gone.
	names := map[string]bool{}
	for _, s := range view.ActiveSchemas(nil) {
		names[s.Name] = true
	}
	if !names["read_file"] {
		t.Fatalf("read_file should be advertised, got %v", names)
	}
	if names["write_file"] {
		t.Fatalf("write_file should be hidden (not read-only), got %v", names)
	}
	if names["bash"] {
		t.Fatalf("bash should be hidden (not read-only), got %v", names)
	}
	if len(names) != 1 {
		t.Fatalf("only read_file should survive, got %v", names)
	}

	// read_file dispatches through to the inner registry.
	ok, err := view.Dispatch(context.Background(), ToolCall{
		ID:    "1",
		Name:  "read_file",
		Input: json.RawMessage(`{}`),
	}, nil)
	if err != nil {
		t.Fatalf("read_file dispatch returned error: %v", err)
	}
	if ok.Output.Kind != ToolOutputSuccess {
		t.Fatalf("read_file should succeed through inner, got %v", ok.Output.Kind)
	}

	// write_file is blocked with a RECOVERABLE error and never reaches inner.
	blocked, err := view.Dispatch(context.Background(), ToolCall{
		ID:    "2",
		Name:  "write_file",
		Input: json.RawMessage(`{}`),
	}, nil)
	if err != nil {
		t.Fatalf("blocked dispatch should return a recoverable ToolOutput, not a Go error: %v", err)
	}
	if blocked.Output.Kind != ToolOutputError {
		t.Fatalf("write_file should be blocked with an error ToolOutput, got %v", blocked.Output.Kind)
	}
	if !blocked.Output.Recoverable {
		t.Fatalf("blocked write_file error must be RECOVERABLE, got fatal")
	}
	if blocked.CallID != "2" {
		t.Fatalf("blocked result must echo the call id, got %q", blocked.CallID)
	}

	// The inner registry only ever saw read_file; write_file never reached it.
	if len(inner.dispatched) != 1 || inner.dispatched[0] != "read_file" {
		t.Fatalf("inner should have seen only [read_file], got %v", inner.dispatched)
	}
}
